modules/schema/src/main/java/org/apache/ignite/internal/schema/BinaryTuplePrefix.java

```diff
@@ -73,6 +97,102 @@ public static BinaryTuplePrefix fromBinaryTuple(BinaryTuple tuple) {
         return new BinaryTuplePrefix(tuple.elementCount(), prefixBuffer);
     }
 
+    private static BinaryTuplePrefix expandTuple(int size, BinaryTuple tuple) {
+        assert size > tuple.elementCount();
+
+        var stats = new Sink() {
+            int dataBeginOffset = 0;
+            int dataEndOffset = 0;
+
+            @Override
+            public void nextElement(int index, int begin, int end) {
+                if (index == 0) {
+                    dataBeginOffset = begin;
+                }
+
+                dataEndOffset = end;
+            }
+        };
+
+        tuple.parse(stats);
+
+        ByteBuffer tupleBuffer = tuple.byteBuffer();
+
+        byte flags = tupleBuffer.get(0);
+        int entrySize = BinaryTupleCommon.flagsToEntrySize(flags);
+
+        ByteBuffer prefixBuffer = ByteBuffer.allocate(
+                        tupleBuffer.remaining()
+                                + (entrySize * (size - tuple.elementCount()))
+                                + Integer.BYTES)
+                .order(ORDER)
+                .put(tupleBuffer.slice().limit(stats.dataBeginOffset)); // header
+
+        int payloadEndPosition = stats.dataEndOffset - stats.dataBeginOffset;
+        for (int idx = tuple.elementCount(); idx < size; idx++) {
+            switch (entrySize) {
+                case Byte.BYTES:
+                    prefixBuffer.put((byte) payloadEndPosition);
+                    break;
+                case Short.BYTES:
+                    prefixBuffer.putShort((short) payloadEndPosition);
+                    break;
+                case Integer.BYTES:
+                    prefixBuffer.putInt(payloadEndPosition);
+                    break;
+                default:
+                    assert false;
+            }
+        }
+
+        prefixBuffer
+                .put(tupleBuffer.slice().position(stats.dataBeginOffset).limit(stats.dataEndOffset)) // payload
+                .putInt(tuple.elementCount())
+                .flip();
+
+        prefixBuffer.put(0, (byte) (flags | PREFIX_FLAG));
+
+        return new BinaryTuplePrefix(size, prefixBuffer);
+    }
+
+    private static BinaryTuplePrefix truncateTuple(int size, BinaryTuple tuple) {
+        assert size < tuple.elementCount();
+
+        var stats = new Sink() {
+            int dataBeginOffset = 0;
+            int dataEndOffset = 0;
+
+            @Override
+            public void nextElement(int index, int begin, int end) {
+                if (index == 0) {
+                    dataBeginOffset = begin;
+                }
+
+                if (index < size) {
+                    dataEndOffset = end;
+                }
+            }
+        };
+
+        tuple.parse(stats);
```

Comment:
This doesn't look optimal to me. Why do we have to iterate over the whole tuple just to get the start and end of the data range? I think we can use a different API (this is pseudocode):
```
tuple.fetch(0, (index, begin, end) -> dataBeginOffset = begin);
tuple.fetch(size - 1, (index, begin, end) -> dataBeginOffset = end);
```

Reply:
Or maybe a better approach is to introduce a new API to the BinaryTuple that will allow to get a "sub-tuple" by providing two indexes

Reply:
great advice with `fetch`, thanks!

about new API: I agree that this looks expensive to create all these byte arrays, but I doubt new api will improve performance in any way. All created objects have very short lifetime and never escape the thread, thus, I believe, they will be created in TLAB, which is relatively cheap

Reply:
About new API: I meant this not only in terms of performance but also in terms of readability. Using `fetch` to find the boundaries of a subarray looks like a battle against the current API. A more straightforward API can be both faster and easier to use

Reply:
> A more straightforward API can be both faster and easier to use

I agree with this, but not sure how such API may look like. Primary cases I interested in are `entireTuple` and `expandTuple`.  I think, former is already good enough. But how to handle `expandTuple`? We basically need to have an ability to append arbitrary amount of null elements to the existing tuple. Beside, we have to reserve additional 4 bytes to append actual number of elements

Reply:
Yeah, looking at `expandTuple` we need the offset values, this sucks. Are you sure that `expandTuple` will be used anywhere?

Reply:
it's used in this very patch. We have BinaryTuple that represents search row. This tuple contains only meaningful values, thus its number of elements may be less than number of indexed fields. 

If you think this method is not very useful as general purpose utility for tuple prefix, then I will move it to sql domain

Reply:
ok, makes sense, let's leave it as-is